src/monitor/group_state_machine.c

```diff
@@ -295,14 +341,533 @@ ProceedGroupState(AutoFailoverNode *activeNode)
+ */
+static bool
+ProceedGroupStateForMSFailover(AutoFailoverNode *activeNode,
+							   AutoFailoverNode *primaryNode)
+{
+	// AutoFailoverFormation *formation = GetFormation(activeNode->formationId);
+	List *standbyNodesGroupList = AutoFailoverOtherNodesList(primaryNode);
+	AutoFailoverNode *candidateNode = NULL;
+	ListCell *nodeCell = NULL;
+
+	int candidateCount = 0;
+	int reportedLSNCount = 0;
+
+	/*
+	 * Done with the single standby code path, now we have several standby
+	 * nodes that might all be candidate for failover, or just some of them.
+	 * First, have them all report the most recent LSN they managed to receive.
+	 */
+	reportedLSNCount = 0;
+
+	foreach(nodeCell, standbyNodesGroupList)
+	{
+		AutoFailoverNode *node = (AutoFailoverNode *) lfirst(nodeCell);
+
+		if (node == NULL)
+		{
+			/* shouldn't happen */
+			ereport(ERROR, (errmsg("BUG: node is NULL")));
+			continue;
+		}
+
+		/* we might have a failover ongoing already */
+		if (IsBeingPromoted(node))
+		{
+			candidateNode = node;
+
+			elog(LOG, "Found candidate node %d (%s:%d)",
+				 node->nodeId, node->nodeName, node->nodePort);
+			continue;
+		}
+
+		/*
+		 * Skip nodes that are not failover candidates (not in SECONDARY or
+		 * REPORT_LSN state).
+		 *
+		 * XXX: what about goalState instead? or in addition to?
+		 */
+		if (node->reportedState != REPLICATION_STATE_SECONDARY
+			&& node->reportedState != REPLICATION_STATE_REPORT_LSN)
+		{
+			continue;
+		}
+
+		/* count how many standby nodes have reached REPORT_LSN */
+		if (IsCurrentState(node, REPLICATION_STATE_REPORT_LSN))
+		{
+			++candidateCount;
+			++reportedLSNCount;
+		}
+		else if (node->goalState != REPLICATION_STATE_REPORT_LSN)
+		{
+			char message[BUFSIZE];
+
+			++candidateCount;
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state of %s:%d to report_lsn "
+				"to find the failover candidate",
+				node->nodeName, node->nodePort);
+
+			AssignGoalState(node, REPLICATION_STATE_REPORT_LSN, message);
+		}
+	}
+
+	/*
+	 * If we can failover, make sure the primary is being demoted before doing
+	 * anything else.
+	 */
+	if (candidateNode != NULL || candidateCount > 0)
+	{
+		/* shut down the primary, known unhealthy (see pre-conditions) */
+		if (IsInPrimaryState(primaryNode))
+		{
+			char message[BUFSIZE];
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state %s:%d to draining after it became unhealthy.",
+				primaryNode->nodeName, primaryNode->nodePort);
+
+			AssignGoalState(primaryNode, REPLICATION_STATE_DRAINING, message);
+		}
+	}
+
+	/*
+	 * If a failover is in progress, continue driving it.
+	 *
+	 * Here we have choosen a failover candidate, which is either being
+	 * promoted to being the new primary (when it already had all the most
+	 * recent WAL, or is done fetching them), or is fetching the most recent
+	 * WAL it's still missing from another standby node.
+	 */
+	if (candidateNode != NULL)
+	{
+		/*
```

Comment:
This code block always returns. Might help readability to move it into a separate function. I keep going back and forth between the top and bottom of this function.

Reply:
Done in a new function!